File: packages/jupyterlab-chat/src/ws-messages.ts
/*
 * Copyright (c) Jupyter Development Team.
 * Distributed under the terms of the Modified BSD License.
 */

import { IUser } from '@jupyter/chat';

/**
 * Typed schema for the per-chat `/api/chat/ws` WebSocket protocol.
 *
 * Every frame is discriminated by two fields:
 * - `type`: the *direction* -- `'client'` (web client -> server) or `'server'`
 *   (server -> web client).
 * - `action`: the specific message within a direction.
 *
 * ```
 * ChatWsMessage       = ClientChatWsMessage | ServerChatWsMessage
 * ClientChatWsMessage = IClientSendMessage | IClientEditMessage
 * ServerChatWsMessage = IServerConnectionMessage | IServerMessageMessage
 *                     | IServerUsersMessage | IServerMetadataMessage
 *                     | IServerWritingMessage
 * ```
 *
 * This mirrors the backend definition in
 * `python/jupyterlab-chat/jupyterlab_chat/ws_messages.py`; the two must stay in
 * sync.
 */

/** Direction discriminator values (the `type` field). */
export const CLIENT = 'client';
export const SERVER = 'server';

/** A raw message object as it travels on the wire (pre-resolution). */
export interface IWireMessage {
  id: string;
  sender: string;
  body?: string;
  time?: number;
  type?: string;
  raw_time?: boolean;
  edited?: boolean;
  deleted?: boolean;
  metadata?: Record<string, any>;
  mime_model?: any;
  attachments?: any[];
  mentions?: string[];
}

// ---------------------------------------------------------------------------
// Client -> server
// ---------------------------------------------------------------------------
export interface IClientSendMessage {
  type: typeof CLIENT;
  action: 'send';
  id: string;
  body: string;
  mentions?: string[];
  metadata?: Record<string, any>;
  attachments?: any[];
  mime_model?: any;
}

export interface IClientEditMessage {
  type: typeof CLIENT;
  action: 'edit';
  id: string;
  body?: string;
  deleted?: boolean;
  edited?: boolean;
  mentions?: string[];
  metadata?: Record<string, any>;
  attachments?: any[];
}

export type IClientChatWsMessage = IClientSendMessage | IClientEditMessage;

// ---------------------------------------------------------------------------
// Server -> client
// ---------------------------------------------------------------------------
export interface IServerConnectionMessage {
  type: typeof SERVER;
  action: 'connection';
  client_id: string;
  id: string;
  user: IUser;
  messages: IWireMessage[];
  users: Record<string, IUser>;
}

export interface IServerMessageMessage {
  type: typeof SERVER;
  action: 'message';
  message: IWireMessage;
}

export interface IServerUsersMessage {
  type: typeof SERVER;
  action: 'users';
  users: Record<string, IUser>;
}

export interface IServerMetadataMessage {
  type: typeof SERVER;
  action: 'metadata';
  metadata: Record<string, any>;
}

export interface IServerWritingMessage {
  type: typeof SERVER;
  action: 'writing';
  user: IUser;
  state: boolean;
  messageID?: string;
  typingIndicator?: string;
}

export type IServerChatWsMessage =
  | IServerConnectionMessage
  | IServerMessageMessage
  | IServerUsersMessage
  | IServerMetadataMessage
  | IServerWritingMessage;

export type IChatWsMessage = IClientChatWsMessage | IServerChatWsMessage;

File: python/jupyterlab-chat/jupyterlab_chat/ws_messages.py
# Copyright (c) Jupyter Development Team.
# Distributed under the terms of the Modified BSD License.

"""Typed schema for the per-chat ``/api/chat/ws`` WebSocket protocol.

Every frame exchanged on the connection is one of the dataclasses defined here,
discriminated by two fields:

* ``type`` -- the *direction*: ``"client"`` (web client -> server) or
  ``"server"`` (server -> web client).
* ``action`` -- the specific message within a direction.

::

    ChatWsMessage    = ClientChatWsMessage | ServerChatWsMessage

    ClientChatWsMessage = ClientSendMessage      # action="send"
                        | ClientEditMessage      # action="edit"

    ServerChatWsMessage = ServerConnectionMessage  # action="connection"
                        | ServerMessageMessage     # action="message"
                        | ServerUsersMessage       # action="users"
                        | ServerMetadataMessage    # action="metadata"
                        | ServerWritingMessage     # action="writing"

The frontend mirrors this union in ``packages/jupyterlab-chat/src/ws-messages.ts``;
the two definitions must stay in sync.

Server frames are constructed as dataclasses and serialized with :func:`to_wire`
(which drops ``None`` fields so optional keys are simply absent on the wire).
Client frames are validated and narrowed by :func:`parse_client_message`.
"""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field
from typing import Any, Dict, List, Literal, Optional, Union

#: Direction discriminator values (the ``type`` field).
CLIENT: Literal["client"] = "client"
SERVER: Literal["server"] = "server"


# ---------------------------------------------------------------------------
# Client -> server
# ---------------------------------------------------------------------------
@dataclass
class ClientSendMessage:
    """A web client asks the server to append a new message."""

    id: str
    body: str = ""
    mentions: List[str] = field(default_factory=list)
    metadata: Optional[Dict[str, Any]] = None
    attachments: Optional[List[dict]] = None
    mime_model: Optional[dict] = None
    type: Literal["client"] = CLIENT
    action: Literal["send"] = "send"


@dataclass
class ClientEditMessage:
    """A web client asks the server to edit (or delete) an existing message."""

    id: str
    body: Optional[str] = None
    deleted: Optional[bool] = None
    edited: Optional[bool] = None
    mentions: Optional[List[str]] = None
    metadata: Optional[Dict[str, Any]] = None
    attachments: Optional[List[dict]] = None
    type: Literal["client"] = CLIENT
    action: Literal["edit"] = "edit"


ClientChatWsMessage = Union[ClientSendMessage, ClientEditMessage]


# ---------------------------------------------------------------------------
# Server -> client
# ---------------------------------------------------------------------------
@dataclass
class ServerConnectionMessage:
    """First frame sent to a client: its id, the chat id, its (server-owned)
    identity, the full message history, and the current users map."""

    client_id: str
    id: str
    user: Dict[str, Any]
    messages: List[dict]
    users: Dict[str, dict]
    type: Literal["server"] = SERVER
    action: Literal["connection"] = "connection"


@dataclass
class ServerMessageMessage:
    """A new or updated message (already attachment-resolved)."""

    message: Dict[str, Any]
    type: Literal["server"] = SERVER
    action: Literal["message"] = "message"


@dataclass
class ServerUsersMessage:
    """An add/update to the users map. Clients merge it into their local map."""

    users: Dict[str, dict]
    type: Literal["server"] = SERVER
    action: Literal["users"] = "users"


@dataclass
class ServerMetadataMessage:
    """An add/update to the chat metadata map. Clients merge it in."""

    metadata: Dict[str, Any]
    type: Literal["server"] = SERVER
    action: Literal["metadata"] = "metadata"


@dataclass
class ServerWritingMessage:
    """An ephemeral writing/typing status pushed by the server (e.g. an AI
    persona). Not persisted to the ``.chat`` file."""

    user: Dict[str, Any]
    state: bool
    messageID: Optional[str] = None
    typingIndicator: Optional[str] = None
    type: Literal["server"] = SERVER
    action: Literal["writing"] = "writing"


ServerChatWsMessage = Union[
    ServerConnectionMessage,
    ServerMessageMessage,
    ServerUsersMessage,
    ServerMetadataMessage,
    ServerWritingMessage,
]

ChatWsMessage = Union[ClientChatWsMessage, ServerChatWsMessage]


# ---------------------------------------------------------------------------
# Serialization / parsing
# ---------------------------------------------------------------------------
def to_wire(message: ServerChatWsMessage) -> str:
    """Serialize a server message to a JSON string, omitting ``None`` fields so
    optional keys are simply absent on the wire."""
    payload = {k: v for k, v in asdict(message).items() if v is not None}
    return json.dumps(payload)


def parse_client_message(data: Any) -> Optional[ClientChatWsMessage]:
    """Validate and narrow a decoded client frame.

    Returns the typed message, or ``None`` if ``data`` is not a well-formed
    client frame (the caller logs and ignores it).
    """
    if not isinstance(data, dict) or data.get("type") != CLIENT:
        return None
    msg_id = data.get("id")
    if not isinstance(msg_id, str) or not msg_id:
        return None

    action = data.get("action")
    if action == "send":
        return ClientSendMessage(
            id=msg_id,
            body=data.get("body", "") or "",
            mentions=list(data.get("mentions") or []),
            metadata=data.get("metadata"),
            attachments=data.get("attachments"),
            mime_model=data.get("mime_model"),
        )
    if action == "edit":
        return ClientEditMessage(
            id=msg_id,
            body=data.get("body"),
            deleted=data.get("deleted"),
            edited=data.get("edited"),
            mentions=data.get("mentions"),
            metadata=data.get("metadata"),
            attachments=data.get("attachments"),
        )
    return None
